File: day3.js
/// Utility functions
const numSort = (a, b) => a - b;
const sum = (a, b) => a + b;

function commonChars(A, B) {
	B = Array.isArray(B) ? B : [B];
	let allCommon = [];
	Array.from(A).forEach((c) => {
		let common = [];
		B.forEach((b) => {
			if (b.indexOf(c) >= 0) {
				common.push(c);
			}
		})
		if (common.length == B.length) {
			allCommon.push(c);
		}
	});
	return allCommon;
}
///

let {sample, input} = require('./input3.js');

function priority(c) {
	let offset = c < 'a' ? 38 : 96;
	return  c.charCodeAt(0) - offset;
}

function part1(input) {
	let rows = input.trim().split("\n");
	let total = 0;
	let sacks = [];

	for (let i=0; i < rows.length; i++) {
		row = rows[i];
		let len = row.length;
		let left = row.slice(0,len/2),
			right = row.slice(len/2);
		let commonVal = 0;

		let common = commonChars(left, [right]);
		commonVal = priority(common[0]);

		//console.log('Row', row, left, right, ' Commmon:', common, ' Priority:', commonVal);
		total += commonVal;
	}
	return total;
}

function part2(input) {
	let rows = input.trim().split("\n");
	let total = 0;
	let groups = [];

	for (let i=0; i < rows.length; i++) {
		row = rows[i];
		let len = row.length;
		//let left = row.slice(0,len/2),
		//	right = row.slice(len/2);
		let common = [];
		let commonVal = 0;

		if (i % 3 == 2) {
			let a = rows[i-2],
				b = rows[i-1],
				c = rows[i];
			groups.push([a, b, c]);
			common = commonChars(a, [b, c]);
			commonVal = priority(common[0]);
			//console.log('Group', [a, b, c], ' Commmon:', common, ' Priority:', commonVal);
			total += commonVal;
		}
	}
	return total;
}

console.log(part1(sample));
console.log(part1(input));
console.log(part2(sample));
console.log(part2(input));

/*

let data = parse(sample);
data.sort((a, b) => a.total - b.total);
console.log(data);
let totals = data.map(e => e.total);
let result = totals.slice(-3);

//totals.sort(numSort);
*/
//console.log(result);
//console.log(result.reduce(sum, 0));

File: day3.exs
defmodule Aoc do
	# Breaks input into groups. Returns list of groups, where each group is a list of integers
	def parseGroups(input) do
		input
		|> String.trim
		|> String.split("\n")
	end

	def priority(c) when c < "a" do
		:binary.first(c) - 38
	end
	def priority(c) do
		:binary.first(c) - 96
	end

	# Given two strings, find the common characters and returns them as a string
	def commonChars(a, b) do
		MapSet.intersection(MapSet.new(String.graphemes(a)), MapSet.new(String.graphemes(b)))
		|> MapSet.to_list
		|> Enum.join("")
	end

	def part1(input) do
		parseGroups(input)
		|> Enum.map(fn sack ->
			{left, right} = String.split_at(sack, div(String.length(sack), 2))
			common = commonChars(left, right)
			priority(common)
		end)
		|> Enum.sum
	end

	def part2(input) do
		parseGroups(input)
		|> Enum.chunk_every(3)
		|> Enum.map(fn ([a, b, c]) ->
			commonChars(a, b)
			|> commonChars(c)
			|> priority
		end)
		|> Enum.sum
	end

	def run(inputFile) do
		IO.puts(["Reading ", inputFile])
		{:ok, input} = File.read(inputFile)
		IO.inspect ["Part 1:", part1(input)]
		IO.inspect ["Part 2:", part2(input)]
	end
end

Aoc.run("sample.3")
Aoc.run("input.3")
